HTML API: Add docs and tests about lowercasing in class_list

sirreal opened the pull request:
The class_list method yields class name strings in lower case. This is
intended but undocumented. Document and add tests for this behavior.




Trac ticket: https://core.trac.wordpress.org/ticket/61520

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
Add docs and tests about lowercasing in class_list

The class_list method yields class name strings in lower case. This is
intended but undocumented. Document and add tests for this behavior.

itzmekhokan commented:
@sirreal changes LGTM.

sirreal commented:
@dmsnell Will you review these documentation changes? These would be good to backport to 6.6.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1694,13 +1694,13 @@ public function has_class( $wanted_class ) {
 
 	/**
-	 * Generator for a foreach loop to step through each class name for the matched tag.
+	 * Generator that yields each class name for the matched tag.
```
Hm. the wording here was carefully chosen to clue people in on how this function is intended to be used. While I understand that `yield` is technically what's occurring on the low level of PHP details, but do you find that this change better leads people to understand how to use the method?

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1694,13 +1694,13 @@ public function has_class( $wanted_class ) {
 	 * Example:
 	 *
-	 *     $p = WP_HTML_Processor::create_fragment( "<div class='free &lt;egg&lt;\tlang-en'>" );
+	 *     $p = new WP_HTML_Tag_Processor( "<div class='FREE &lt;egg&lt;\tlang-en'>" );
```
This bit of the change was wrong:

```suggestion
	 *     $p = WP_HTML_Processor::create_fragment( "<div class='FREE &lt;egg&lt;\tlang-en'>" );
```

sirreal committed:
Fix wrong constructor in processor doc

sirreal committed:
Add minimal note about ASCII-lowercase class names

Revert other documentation tweaks and only include changes to reflect
the ASCII-lowercase behavior.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1694,13 +1694,13 @@ public function has_class( $wanted_class ) {
+	 * Generator that yields each class name for the matched tag.
 	 *
-	 * This generator function is designed to be used inside a "foreach" loop.
+	 * Class names will be lowercased and unique class names will be yielded once.
```
I'm really uncomfortable both taking out relevant information for consumers of this method and then adding new kinds of specifications for things that aren't specified.

In particular I've attempted to avoid using the term lower-case and upper-case except where necessary. if you can find some notion in the CSS specification indicating that this is _part_ of what it means to be a CSS class, then I think it's appropriate. I'm far less versed in the CSS spec so I've been reluctant to add constraints that aren't there.

To this end I've favored creating our own term to avoid the ambiguity, which is `$comparable_name`.

The best I can find is in [CSS2](https://www.w3.org/TR/CSS2/syndata.html#characters) (repeated up through [CSS Selectors, Level 4](https://www.w3.org/TR/selectors-4/#case-sensitive)), which states

> All CSS syntax is case-insensitive within the ASCII range…

and

> All Selector syntax is ASCII case-insensitive…

The most confusing bit, which is what I was referring to earlier when we discussed this, is a note about quirks mode. In [6.6 Class selectors](https://www.w3.org/TR/selectors-4/#class-html) it says:

> When matching against a document which is in [quirks mode](https://dom.spec.whatwg.org/#concept-document-quirks), class names must be matched [ASCII case-insensitively](https://infra.spec.whatwg.org/#ascii-case-insensitive); class selectors are otherwise case-sensitive, only matching class names they are [identical to](https://infra.spec.whatwg.org/#string-is).

We obviously can't know about quirks mode in the Tag Processor, and I don't even think we want to, but I think this section is still speaking specifically about the `class` attribute and not the CSS classes.

Funny thing is, I can't find any mention of what a class _is_. All I can find in the CSS specifications are discussions about _class selectors_.

The [DOM specification](https://dom.spec.whatwg.org/#dom-element-classlist) indicates that the `classList getter` is a token set whose token list object as known as "the element's *classes*."

It further hearkens back to the CSS spec [when saying](https://dom.spec.whatwg.org/#concept-getelementsbyclassname),

> The comparisons for the classes must be done in an ASCII case-insensitive manner if _root's_ node document's mode is "quirks"; otherwise in an identical to manner.

I don't think there's a specification on `classList` forcing it to be lower-case.
Safari/WebKit seems to do it by happenstance

Hang on, we're continuing this chat in person and you can ignore all this. it's partial and possibly out-dated.

dmsnell replied:
| No-Quirks Mode | Quirks Mode |
|---|---|
| <img width="603" alt="Screenshot 2024-06-28 at 1 18 12 PM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/b5853c16-e86c-475b-91df-1f0dfac25121"> | <img width="605" alt="Screenshot 2024-06-28 at 1 18 36 PM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/b1c58e5a-bd1b-4a61-815f-6341c50e1b59"> |

sirreal commented:
This may not be the best approach. Rather than documenting the specifics of the behavior here, it seems best to audit the class-related methods.

dmsnell reviewed (commented):
Left some notes, but we talked in person and I think it's worth ignoring all I wrote except for historic purposes.
